### PR TITLE
support unix-abstract schema

### PR DESCRIPTION
**Pull request by resec**
Resolving #3998

### CODE REVIEW

**Comment by resec** [comment]
Needs help here, the "retry" test is passing on my ubuntu machine with go 1.15.6, yet it failed here.

Please point out some directions I should investigate? and how can I re-trigger the pr testing? Thanks..

File: internal/grpcutil/target_test.go [hunk 1]
```diff
@@ -84,6 +84,14 @@ func TestParseTargetString(t *testing.T) {
 		{targetStr: "unix:/a/b/c", want: resolver.Target{Scheme: "unix", Authority: "", Endpoint: "/a/b/c"}, wantWithDialer: resolver.Target{Scheme: "", Authority: "", Endpoint: "unix:/a/b/c"}},
 		{targetStr: "unix:///a/b/c", want: resolver.Target{Scheme: "unix", Authority: "", Endpoint: "/a/b/c"}},
 
+		{targetStr: "unix-abstract:a/b/c", want: resolver.Target{Scheme: "unix-abstract", Authority: "", Endpoint: "a/b/c"}},
```
Comment (dfawley) [inline internal/grpcutil/target_test.go:87]: `unix-abstract:///a/b/c` and `unix-abstract://authority/a/b/c` should parse as expected as well.  I don't believe they will with the changes in this PR currently.

File: internal/resolver/unix/unix.go [hunk 2]
```diff
@@ -26,20 +26,28 @@ import (
 	"google.golang.org/grpc/resolver"
 )
 
-const scheme = "unix"
+const unixScheme = "unix"
+const unixAbstractScheme = "unix-abstract"
 
-type builder struct{}
+type builder struct {
+	scheme string
```
Comment (dfawley) [inline internal/resolver/unix/unix.go:33]: Can we use `target.Scheme` to determine whether to prepend the null byte instead of needing this?

File: internal/transport/http_util.go [hunk 3]
```diff
@@ -603,6 +603,11 @@ func newFramer(conn net.Conn, writeBufferSize, readBufferSize int, maxHeaderList
 // parseDialTarget returns the network and address to pass to dialer.
 func parseDialTarget(target string) (string, string) {
 	net := "tcp"
+	if strings.HasPrefix(target, "unix-abstract:") {
+		// handle unix-abstract:addr which will fail with url.Parse
+		// and need prepend "\x00" to the parsed address
+		return "unix", "\x00" + strings.SplitN(target, ":", 2)[1]
+	}
```
Comment (dfawley) [inline internal/transport/http_util.go:610]: This should not be necessary.  We only do this for "unix" for backward compatibility reasons.  I.e. we don't need to support `passthrough:///unix-abstract:path`.

**Review by dfawley — CHANGES_REQUESTED** [review]
Thanks for the PR.  The failing test looked like a flake, so I re-ran everything.  A few comments inline.

File: internal/resolver/unix/unix.go (hunk 2 above)
Comment (resec) [reply internal/resolver/unix/unix.go:33]: It seems that we have to store the schema name inside the builder so `builder.Scheme()` can return the correct one.

File: internal/transport/http_util.go (hunk 3 above)
Comment (resec) [reply internal/transport/http_util.go]: Understood, removed along with the test cases

File: internal/grpcutil/target_test.go (hunk 1 above)
Comment (resec) [reply internal/grpcutil/target_test.go:87]: Test cases added

File: internal/resolver/unix/unix.go (hunk 2 above)
Comment (dfawley) [reply internal/resolver/unix/unix.go:33]: Oh yes, I missed that, sorry.

File: internal/grpcutil/target_test.go [hunk 4]
```diff
@@ -91,6 +91,8 @@ func TestParseTargetString(t *testing.T) {
 		{targetStr: "unix-abstract:/ a///://::!@#$%^&*()b", want: resolver.Target{Scheme: "unix-abstract", Authority: "", Endpoint: "/ a///://::!@#$%^&*()b"}},
 		{targetStr: "unix-abstract:passthrough:abc", want: resolver.Target{Scheme: "unix-abstract", Authority: "", Endpoint: "passthrough:abc"}},
 		{targetStr: "unix-abstract:unix:///abc", want: resolver.Target{Scheme: "unix-abstract", Authority: "", Endpoint: "unix:///abc"}},
+		{targetStr: "unix-abstract:///a/b/c", want: resolver.Target{Scheme: "unix-abstract", Authority: "", Endpoint: "///a/b/c"}},
+		{targetStr: "unix-abstract://authority/a/b/c", want: resolver.Target{Scheme: "unix-abstract", Authority: "", Endpoint: "//authority/a/b/c"}},
```
Comment (dfawley) [inline internal/grpcutil/target_test.go:95]: This target string *should* parse as `Authority: "authority", Endpoint: "/a/b/c"`

Comment (resec) [reply internal/grpcutil/target_test.go:95]: The `unix-abstract` schema has no authority by the definition in https://github.com/grpc/grpc/blob/master/doc/naming.md, same as `unix` schema.

Am I missing something?

Comment (dfawley) [reply internal/grpcutil/target_test.go:95]: Theoretically, *all* schemes should use the exact same logic to parse a URI into its components.  We unfortunately have some hackery around here for backward compatibility reasons.

It's illegal for an authority to be specified for `unix-abstract`, but that should make the name resolver fail the attempt to include one, not parse it as part of the endpoint.

Comment (resec) [reply internal/grpcutil/target_test.go:95]: Understood, if this is for backward compatibility reasons, I am fine with it.

Just one final confirmation, if we do parse the authority, we are actually compromising the namespace of `unix-abstract`? Because it is legal by definition that user can use any bytes as the abstract socket name, yet our parser blocks it.

Comment (dfawley) [reply internal/grpcutil/target_test.go:95]: > Understood, if this is for backward compatibility reasons, I am fine with it.

Not quite.  There is some hackery in parsing existing schemes (e.g. `unix`) for backward compatibility.  I want `unix-abstract` to be parsed the way all target strings are supposed to be parsed.

> Just one final confirmation, if we do parse the authority, we are actually compromising the namespace of unix-abstract? Because it is legal by definition that user can use any bytes as the abstract socket name, yet our parser blocks it.

We aren't compromising anything; this is how Java and C do their parsing here as well.  The socket name *can* be any sequence of bytes, but they need to be specified correctly.  I.e. if you want the endpoint to be "//a/b/c" for some reason, then you will need to use `"unix-abstract://///a/b/c"` because we don't currently support percent-encoding; otherwise you could do `"unix-abstract:%2F%2Fa%2Fb%2Fc"` to avoid making the `"//a/"` appear like it's specifying an authority of `a` and an endpoint of "/b/c".

Basically we really want RFC 3986-compliant URI parsing, but we have to grandfather in uses from when our parsing was ad-hoc and wrong in some cases.  For new schemes, we probably could use `url.Parse` but let's push off such a major change for another time.

Comment (resec) [reply internal/grpcutil/target_test.go]: Understood, thanks for your detailed explanation. Already added the code for authority parsing for `unix-abstract` schema

Also added two more test cases for `unix-abstract` schema:
- `unix-abstract:///` parsed as empty Authority and Endpoint `/`
- `unix-abstract://authority` parsed as empty Authority and Endpoint `//authority`

Let me know if they are unexpected. Thanks for your time.

File: internal/grpcutil/target.go [hunk 5]
```diff
@@ -41,29 +41,52 @@ func split2(s, sep string) (string, string, bool) {
 // not parse "unix:[path]" cases. This should be true in cases where a custom
 // dialer is present, to prevent a behavior change.
 //
-// If target is not a valid scheme://authority/endpoint, it returns {Endpoint:
-// target}.
+// If target is not valid dns/unix/unix-abstract schemas as specified in
+// https://github.com/grpc/grpc/blob/master/doc/naming.md,
```
Comment (dfawley) [inline internal/grpcutil/target.go:45]: This isn't quite accurate.  We support custom schemes, and those will parse into scheme/authority/endpoint as well.  I think you can leave the old comment how it was, just add the reference link.

File: internal/grpcutil/target.go [hunk 6]
```diff
@@ -41,29 +41,52 @@ func split2(s, sep string) (string, string, bool) {
...
-	ret.Scheme, ret.Endpoint, ok = split2(target, "://")
-	if !ok {
-		if strings.HasPrefix(target, "unix:") && !skipUnixColonParsing {
-			// Handle the "unix:[local/path]" and "unix:[/absolute/path]" cases,
-			// because splitting on :// only handles the
-			// "unix://[/absolute/path]" case. Only handle if the dialer is nil,
-			// to avoid a behavior change with custom dialers.
-			return resolver.Target{Scheme: "unix", Endpoint: target[len("unix:"):]}
+	switch {
+	case strings.HasPrefix(target, "unix-abstract:"):
```
Comment (dfawley) [inline internal/grpcutil/target.go:50]: Please change this to `if`, and `return ret` at the end of the `if`.  Then you can out-dent the `default` and the diffs should be cleaner.  (Plus less-indented code for the "default" flow is preferable when possible.)

File: internal/grpcutil/target_test.go (hunk 4 above)
Comment (dfawley) [reply internal/grpcutil/target_test.go]: > Also added two more test cases for unix-abstract schema:

Those also sound like great test cases, thanks.

**Review by dfawley — CHANGES_REQUESTED** [review]
Looks good now, just a couple minor things.  Thank you!

File: internal/grpcutil/target.go (hunk 6 above)
Comment (resec) [reply internal/grpcutil/target.go]: Understood and done

File: internal/grpcutil/target.go (hunk 5 above)
Comment (resec) [reply internal/grpcutil/target.go]: done

**Review by dfawley — APPROVED** [review]
LGTM; thanks!
